### PR TITLE
docs: why do we not use argo

### PR DESCRIPTION
**Pull request by STRRL**
Signed-off-by: STRRL <str_ruiling@outlook.com>

### What problem does this PR solve?

Rendered markdown: [link](https://github.com/STRRL/chaos-mesh/blob/why-do-we-not-use-argo/workflow/docs/why-do-we-not-use-argo.md)
This doc represents why we choose to build our own workflow engine, but not using argo.

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by STRRL** [comment]
If you have any other questions or ideas, welcome to leave comments!

File: workflow/docs/why-do-we-not-use-argo.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+I think the most advantage is reducing our jobs, we do not need to re-implement another workflow engine. We could build a workflow by writing some scripts with `kubectl`.
+
+## Argo workflow is not the best way to describe declarative chaos experiments
+
+Argo supports "script", "container" and "resource" workloads, it could integrate with Chaos Mesh to apply/recover chaos experiments. But this way to apply chaos experiments is not declarative anymore, it turns to imperative.
+
+Argo is so powerful and it's impossible to ask users always for following the best practices. Users could write complex scripts for determining the situation, then `kubectl apply` or `kubectl patch` something, once a command make unexpected side-effects, it will let whole system unstable.
+
+We define the workflow that could describe chaos experiments in the declarative, changing chaos experiments is just like "from expected status A turning into expected status B", users should not considering about "remove some chaos CR object" or "create new chaos CR object". We also allow users (only) to execute their scripts in `Task` node, it will not affect the changing of chaos experiment states.
```
Comment (AsterNighT) [inline workflow/docs/why-do-we-not-use-argo.md:29]: Is there now any thoughts on a declarative way of defining workflow status? I don't mean to judge, just being curious. You see, I suppose there has to be somewhere to store the status, and we all agree that using a CRD is an idea not that good, not being native to k8s. If possible, I'm looking forward to a further and deeper discussion or explaination on the designing of workflow.

File: workflow/docs/why-do-we-not-use-argo.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+Argo is an amazing workflow engine for common purpose. Argo provides "sequential/parallel execution", "conditional branches", "DAG"; and workloads like "script", "container", "resource".
+
+I think the most advantage is reducing our jobs, we do not need to re-implement another workflow engine. We could build a workflow by writing some scripts with `kubectl`.
+
+## Argo workflow is not the best way to describe declarative chaos experiments
+
+Argo supports "script", "container" and "resource" workloads, it could integrate with Chaos Mesh to apply/recover chaos experiments. But this way to apply chaos experiments is not declarative anymore, it turns to imperative.
+
+Argo is so powerful and it's impossible to ask users always for following the best practices. Users could write complex scripts for determining the situation, then `kubectl apply` or `kubectl patch` something, once a command make unexpected side-effects, it will let whole system unstable.
```
Comment (fewdan) [inline workflow/docs/why-do-we-not-use-argo.md:27]: "the whole system"

File: workflow/docs/why-do-we-not-use-argo.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+## Why the other chaos platform choose to use Argo, but we do not prefer to use it
+
+Actually, for example, litmus chaos is not strictly declarative chaos experiments, its `ChaosExperiments` contains scripts that how to inject chaos. During executing, a `ChaosEngine` will instantiate it into a `Job`, then running those scripts. If something unexpected happened, such as the pod was killed by OOM killer, the dirty thing is just left here without recovering anymore. In other words, it just a wrapper of `Job`, with more effective selectors and conditions. So it is suitable for Argo, it is imperative commands from start to end.
+
+In other words, we do not use Argo, which does not mean we reject Argo. We are trying to defining something that could schedule chaos, and independent with Argo.
+
+## Other misc
+
+Applications based on kubernetes have an issue: it's nearly not possible to deploy different instances with different versions in one cluster. If Chaos Mesh depend on argo version X, and the user already installed argo version Y, it might be incompatible, there is no way to resolve it.
```
Comment (fewdan) [inline workflow/docs/why-do-we-not-use-argo.md:47]: "depends on"

File: workflow/docs/why-do-we-not-use-argo.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Chaos Mesh have the plans for more situation not only kubernetes, just like CaaS(Chaos engineering as a Service). If one day we need to migrate workflow into a non-kubernetes cluster, I think our own workflow is easier to do.
```
Comment (fewdan) [inline workflow/docs/why-do-we-not-use-argo.md:49]: "has the plans"

File: workflow/docs/why-do-we-not-use-argo.md (hunk 1 above)
Comment (STRRL) [reply workflow/docs/why-do-we-not-use-argo.md:29]: > You see, I suppose there has to be somewhere to store the status, and we all agree that using a CRD is an idea not that good, not being native to k8s.

I think using CRD could not be determined as "good" or "bad", CRD is the only way that we store the status by Kubernetes API.

I want to present that "state machine with event-driven" is bad. It has some discussions at https://github.com/chaos-mesh/chaos-mesh/pull/1418#issuecomment-765224545

And I am still trying to implement workflow without state-machine anymore, it is nearly accomplished (on theory). The core is defining the expected status more clearly, makes as more as possible thing have been determined as the object be created.

I will update the design documents after I finish the proving of concepts.

File: workflow/docs/why-do-we-not-use-argo.md (hunk 4 above)
Comment (STRRL) [reply workflow/docs/why-do-we-not-use-argo.md]: all fixed! PTAL

File: workflow/docs/why-do-we-not-use-argo.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (AsterNighT) [reply workflow/docs/why-do-we-not-use-argo.md:29]: That'll be possible with some hack in the current event workflow designing, for example, allow some node to run until it has received some event from its siblings. But that is a bit nasty, I wouldn't say I'm fond of that.

[the remark this answers is not in the document]
Comment (STRRL) [reply workflow/docs/why-do-we-not-use-argo.md:29]: > For example, how do we describe the following combined workflow?

> But I hope our workflow can cover most use cases of the cargo.

I do not think the workflow in Chaos Mesh is a common-usage workflow for now. We do not support running jobs during chaos experiments affected. Chaos Mesh workflow was only designed to schedule "Chaos".

As users could not describe jobs in Chaos Mesh Workflow, users are couraged using another workflow engine like argo that schedules their jobs and just integrates with Chaos Mesh Workflow.

The scenario that you mentioned is very interesting! I think we could have more discussion.

[the remark this answers is not in the document]
Comment (STRRL) [reply workflow/docs/why-do-we-not-use-argo.md:29]: > That'll be possible with some hack in the current event workflow designing, for example, allow some node to run until it has received some event from its siblings.

I think it's necessary that let workflow doing something triggered by some event, not only controller by `duration` and `deadline`. Although it's not be represented in recent codes and docs, I believe we will make it.

**Comment by STRRL** [comment]
PTAL @Hexilee @AsterNighT @fewdan @cwen0 @WangXiangUSTC @YangKeao, this PR only contains documents, PTAL and make your review. 🥺

**Comment by AsterNighT** [comment]
I could use some details in the implementation. Would you kindly explain that briefly, especially on how do you decide to overcome the difficulties mentioned above?

**Comment by STRRL** [comment]
> I could use some details in the implementation. Would you kindly explain that briefly, especially on how do you decide to overcome the difficulties mentioned above?

I do not get your point. 🙈 I think what we need is a declartive workflow that could interactive with Chaos and other concepts(like Probe, the health checker), with some features that could not make it easy with argo.

And what is the "difficulties" you mean? The workflow which combines test and chaos mentioned by @Hexilee? Or something in the docs?

**Review by Hexilee — APPROVED** [review]
LGTM as an explanation, we can discuss the implementation of declarative workflows in latter docs.

**Comment by AsterNighT** [comment]
> > I could use some details in the implementation. Would you kindly explain that briefly, especially on how do you decide to overcome the difficulties mentioned above?
> 
> I do not get your point. 🙈 I think what we need is a declartive workflow that could interactive with Chaos and other concepts(like Probe, the health checker), with some features that could not make it easy with argo.
> 
> And what is the "difficulties" you mean? The workflow which combines test and chaos mentioned by @Hexilee? Or something in the docs?

Something like that. In fact I haven't gotten a clear view of workflow, like how it'll handle problems like avoiding state machine and some sort of event-driven mentioned by @Hexilee. There's no way for me to tell about the topic `why do we not use argo` since I'm not clear about the pros and cons of workflow.

**Comment by STRRL** [comment]
> In fact I haven't gotten a clear view of workflow, like how it'll handle problems like avoiding state machine and some sort of event-driven mentioned by @Hexilee. 

This document does not make any represents about what to do, and how to do it. I think you could get them from the design document of our workflow.

> There's no way for me to tell about the topic `why do we not use argo` since I'm not clear about the pros and cons of workflow.

I think we have no "dedicate" or a "professional" instruction that tell us "do not use argo". We just want a thing that could meet our other requirements, and argo could not 100% satisfy us.

**Review by AsterNighT — APPROVED** [review]
LGTM now.

**Pull request merged**